### ISSUE TITLE
Add support for Webpack resolving

### ISSUE THREAD

**Issue by GreenGremlin** [ternjs/tern#663]
The Webpack resolver is nicely factored out of the core Webpack code ([webpack/enhanced-resolver](https://github.com/webpack/enhanced-resolve)). It looks like it should be feasible to implement a plugin to read a Webpack config file and resolve dependencies using the Webpack resolver plugins.

It looks like there would be some overlap between the Webpack plugins and the Tern plugins (modules, es_modules, etc.). It may be best to bypass those plugins when using Webpack.

See:
  - Webpack resolver - [webpack/enhanced-resolver](https://github.com/webpack/enhanced-resolve)
  - An example of Webpack using the resolver plugins - [webpack/webpack](https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L292)

**Comment by GreenGremlin** [comment]
I may take a stab at this, if I have time and energy. Any advice on what the integration should look like would be greatly appreciated.

**Comment by marijnh** [comment]
You'll want to take inspiration from `plugin/node_resolve.js` -- load the `"modules"` plugin, and then push a function to `server.mod.modules.resolvers` that implements your resolution algorithm.

**Comment by chemzqm** [comment]
Here's my plugin for webpack https://gist.github.com/chemzqm/7ceee3040f1377a534e6 using webpack/enhanced-resolver and changed code from `plugin/node_resolve.js`.
It works great on my webpack project!

**Comment by marijnh** [comment]
Where do WebPack's rules differ from node's rules?

**Comment by chemzqm** [comment]
The main different for me is webpack can understand browser field, it's kind of module name alias https://gist.github.com/defunctzombie/4339901, as the modules I use relay on this field to work. The resolving process could also be configured by the user https://webpack.github.io/docs/resolving.html#resolving-a-module-path it's not well documented yet

**Comment by marijnh** [comment]
@chemzqm I've cleaned up and integrated your plugin in the [webpack branch](https://github.com/ternjs/tern/tree/webpack), but the test I added to verify that it is actually respecting the `"browser"` field is failing. Could you look into that?

**Comment by chemzqm** [comment]
This line https://github.com/ternjs/tern/blob/webpack/plugin/webpack.js#L18
need to add `browser` filed
``` js
new req.DirectoryDescriptionFilePlugin("package.json", ["browser","main"]),
```
would works,  I've tested this on my machine.

It's not for webpack only, other browser bundler like browserify also use the browser field

**Comment by marijnh** [comment]
That does help. I've merged the fixed plugin in ae3ce30. At some point in the future, we might want to move to a single, configurable module resolution plugin based on `enhanced-resolve` (since this is cleaner and more portable than calling into node.js internals, as `node_resolve` does), but for now I don't have the time to look into that.
